gobblin-service/src/main/java/org/apache/gobblin/service/modules/restli/GobblinServiceFlowConfigResourceHandler.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.gobblin.service.modules.scheduler.GobblinServiceJobScheduler;
+import org.apache.gobblin.service.modules.utils.HelixUtils;
+
+
+/**
+ * A high available flow config resource handler which consider the leadership change.
+ * When a non-master status detected, it will forward the rest-li request
+ * to the master node. Otherwise it will handle it locally.
+ */
+@Slf4j
+public class GobblinServiceFlowConfigResourceHandler implements FlowConfigsResourceHandler {
+  @Getter
+  private String serviceName;
+  private boolean flowCatalogLocalCommit;
+  private FlowConfigResourceLocalHandler localHandler;
+  private Optional<HelixManager> helixManager;
+  private GobblinServiceJobScheduler jobScheduler;
+
+  public GobblinServiceFlowConfigResourceHandler(String serviceName, boolean flowCatalogLocalCommit,
+      FlowConfigResourceLocalHandler handler,
+      Optional<HelixManager> manager,
+      GobblinServiceJobScheduler jobScheduler) {
+    this.flowCatalogLocalCommit = flowCatalogLocalCommit;
+    this.serviceName = serviceName;
+    this.localHandler = handler;
+    this.helixManager = manager;
+    this.jobScheduler = jobScheduler;
+  }
+
+  @Override
+  public FlowConfig getFlowConfig(FlowId flowId)
+      throws FlowConfigLoggedException {
+    return this.localHandler.getFlowConfig(flowId);
+  }
+
+  /**
+   * Method to handle create Restli request.
+   * In load balance mode, we will handle flowCatalog I/O locally before forwarding the message to Helix (Active) node.
+   * Please refer to {@link FlowConfigResourceLocalHandler#createFlowConfig(FlowConfig)}. It only handle flowCatalog I/O.
+   *
+   * The listeners of {@link org.apache.gobblin.runtime.spec_catalog.FlowCatalog} won't be triggered in balance mode.
+   */
+  @Override
+  public CreateResponse createFlowConfig(FlowConfig flowConfig)
+      throws FlowConfigLoggedException {
+    String flowName = flowConfig.getId().getFlowName();
+    String flowGroup = flowConfig.getId().getFlowGroup();
+
+    if (this.helixManager.isPresent() && !this.helixManager.get().isConnected()) {
+      // Specs in store will be notified when Scheduler is added as listener to FlowCatalog, so ignore
+      // .. Specs if in cluster mode and Helix is not yet initialized
+      log.warn("System not yet initialized. Skipping flowConfig Addition: " + flowConfig);
+      throw new FlowConfigLoggedException(HttpStatus.S_500_INTERNAL_SERVER_ERROR,
+          "System not yet initialized. Skipping create flowConfig [flowName=" + flowName + " flowGroup=" + flowGroup + "]");
+    }
+
+    try {
+      if (!jobScheduler.isActive() && helixManager.isPresent()) {
+        if (this.flowCatalogLocalCommit) {
+          // We will handle FS I/O locally for load balance before forwarding to remote node.
+          this.localHandler.createFlowConfig(flowConfig, false);
+        }
+
+        HelixUtils.sendUserDefinedMessage(ServiceConfigKeys.HELIX_FLOWSPEC_ADD,
```

Comment:
This block can also be shared. Just pass the messageSubtype and flowConfig as a parameters.